KAFKA-15169: Added TestCase in RemoteIndexCache

iit2009060 opened the pull request:
### **Test Cases Covered** 

1. **Index Files already exist on disk but not in Cache** i.e. RemoteIndexCache should not call  remoteStorageManager to fetch it instead cache it from the local index file present.
2. **RSM returns CorruptedIndex File**  i.e. RemoteIndexCache should throw CorruptedIndexException instead of successfull execution. 
3. **Deleted Suffix Indexes file already present on disk**  i.e. If cleaner thread is slow , then there is a chance of deleted index files present on the disk while in parallel same index Entry is invalidated. To understand more refer https://issues.apache.org/jira/browse/KAFKA-15169

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

iit2009060 committed:
Added TestCase in RemoteIndexCache

divijvaidya commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -541,6 +541,126 @@ class RemoteIndexCacheTest {
   }
 
+  @Test
+  def testOffsetIndexFileAlreadyExistOnDiskButNotInCache(): Unit ={
+    val remoteIndexCacheDir = new File(tpDir,RemoteIndexCache.DIR_NAME)
```
Please follow style guide used in rest of the code. For example, space between = and {.
You can find more information here: https://kafka.apache.org/coding-guide.html 

I remember your previous PR also had similar comments. I would recommend to set your IDE to use this style and automatically format newly added code.

iit2009060 replied:
Thanks , I have enabled on IDE to format it on save everytime.

iit2009060 replied:
done

divijvaidya commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -541,6 +541,126 @@ class RemoteIndexCacheTest {
   }
 
+  @Test
+  def testOffsetIndexFileAlreadyExistOnDiskButNotInCache(): Unit ={
+    val remoteIndexCacheDir = new File(tpDir,RemoteIndexCache.DIR_NAME)
```
please create a getter in the RemoteIndexCache to get this dir, add javadoc for it `// visible for testing` and use that here. It will make this test independent of the remote cache location in case we change that in future,

iit2009060 replied:
done

divijvaidya commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -541,6 +541,126 @@ class RemoteIndexCacheTest {
+      Files.walk(remoteIndexCacheDir.toPath)
+        .filter(Files.isRegularFile(_))
+        .filter(path => path.getFileName.toString.endsWith(suffix))
+        .forEach(f => Files.move(f,f.resolveSibling(f.getFileName().toString().stripSuffix(tempSuffix))))
```
Instead of files.move, please use our custom utility function `Utils.atomicMoveWithFallback`

iit2009060 replied:
done

divijvaidya commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -541,6 +541,126 @@ class RemoteIndexCacheTest {
+    // rsm should not be called again
+    // instead files exist on disk
+    // should be used
+    verifyFetchIndexInvocation(count = 1)
```
should this be 0 if there is going to be no fetch invocation?

iit2009060 replied:
@divijvaidya  To pre-create the valid index file in remote storage cache dir , I initially fetched it from the remote storage and then run the scenario. The 1 count is for pre-processing.

showuon replied:
Thanks for the explanation. So that will confuse readers. Please do either:
1. reset the mock rsm
or
2. run `verifyFetchIndexInvocation(count = 1)` again before L672, so that we can make sure the count is not increasing even we call `cache.getIndexEntry` again.

showuon replied:
@iit2009060 , have you addressed this comment?

iit2009060 replied:
Hey @showuon  Thanks for identifying it , I missed it. 
Changes Done.

divijvaidya commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -541,6 +541,126 @@ class RemoteIndexCacheTest {
+        // The size of the string written in the file is 12 bytes,
+        // but it should be multiple of Offset Index EntrySIZE which is equal to 8.
+        pw.close()
+        val offsetIdx = createOffsetIndexForSegmentMetadata(metadata)
```
there is a bug in existing `createOffsetIndexForSegmentMetadata` and similar function. They create cache at the wrong place in the line:
`new OffsetIndex(remoteOffsetIndexFile(tpDir, metadata)`
Instead of `tpDir`, it should point to remote cache director which is a folder inside `tpDir`.

Please fix that and see if it impacts your test in any way.

iit2009060 replied:
@divijvaidya  But when we call get index entry , it will create indexes file under remote  cache directory only. It will read  file content from tpDir  and create indexes in "remoteIndexCache" dir. 
IMO  "createOffsetIndexForSegmentMetadata" will create indexes file and tpDir act as a placeholder for files fetched from remoteStorage. If we change it to remote cache dir , all test cases will always have the file exist behaviour.

iit2009060 replied:
@divijvaidya  We can keep this function i.e createOffsetIndexForSegmentMetadata to be used for indexes  received from remote storage rather than remoteindexcache. 

The existing test case are working because they directly put values into internalCache(Caeffine) rather than the getIndexEntry route. 

In the above test case I want to corrupt the Index files received from remote storage rather than from remote index cache.
getIndexEntry does following steps 
1. Fetch indexes from remote Storage( i.e indexes stored in tpDir directory as per the mock)
2. Manually Corrupt the indexes file stored in tpDir Directory after fetch from remote storage.
3. It tries to copy indexes files in tpDir directory to remote index cache directory. 
4. Run sanity check on indexes stored in tpDir/remoteindexcache directory. 
5. It throws CorruptedIndexException.

divijvaidya commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -541,6 +541,126 @@ class RemoteIndexCacheTest {
+        // The size of the string written in the file is 12 bytes,
+        // but it should be multiple of Offset Index EntrySIZE which is equal to 8.
+        pw.close()
+        val offsetIdx = createOffsetIndexForSegmentMetadata(metadata)
```
could we have 3 variants of this test where we are corruption all 3 indexes one by one. You can use `@Parameterized` to run those scenarios in same test.

iit2009060 replied:
@divijvaidya Is it possible to parametrize based on function as a parameter  ? Otherwise I need to write if else condition in the same test ?
Corrupting Time and Offset Index is easy based on entry_size , any suggestion how to do it for TransactionIndex ?

showuon replied:
@iit2009060 , if / else is fine with me, just make the test clear.

For txnIndex corruption, I usually check other tests if they did something similar as we want. Here's a good reference:

https://github.com/apache/kafka/blob/fbc39bcdbe5ebcb07435f467b77f4b7df2e910f4/core/src/test/scala/unit/kafka/log/TransactionIndexTest.scala#L62-L73

iit2009060 replied:
@divijvaidya  @showuon done

divijvaidya commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -541,6 +541,126 @@ class RemoteIndexCacheTest {
+    // wait until entry is marked for deletion
+    TestUtils.waitUntilTrue(() => entry.isMarkedForCleanup,
+      "Failed to mark cache entry for cleanup after invalidation")
+    TestUtils.waitUntilTrue(() => entry.isCleanStarted,
```
Clean started doesn't mean that index has been deleted. There could be a case where next line executes even though file has not been deleted yet. I would suggest to wrap the next assertions in `waitUntilTrue()`. It would make the test non-flaky.

iit2009060 replied:
done

iit2009060 committed:
Use waitUntilTrue,correct Indentation

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -541,6 +542,131 @@ class RemoteIndexCacheTest {
+
+    // restore index files
+    renameRemoteCacheIndexFileFromDisk(tempSuffix)
+    // validate cache entry for the above key should be  null
```
nit: additional space before null.

iit2009060 replied:
done

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -541,6 +542,131 @@ class RemoteIndexCacheTest {
   }
 
+  @Test
+  def testIndexFileAlreadyExistOnDiskButNotInCache(): Unit = {
```
Will this situation happen in real case? If so, it should be a bug, right? @divijvaidya , WDYT?

iit2009060 replied:
Yes @showuon  , this can happen in real case
1. Call getIndexEntry 
2. Offset index is successfully cached 
3. But TimeIndex received from remote storage is corrupted 
4. getIndexEntry throws Corrupt Exception
But it leaves the offset index file on disk in remote cache directory.

showuon replied:
Hmm... make sense! So, don't you think we should make the test case go through the above steps 1 to 4, and then verify next getIndexEntry can still work fine?

showuon replied:
Maybe we can merge it into the `testRSMReturnCorruptedIndexFile` test case?

iit2009060 replied:
@showuon  This test   "testIndexFileAlreadyExistOnDiskButNotInCache" run for all indexes where file already exists(manually created). 
I can write  another test where instead of manually we derive the existence  from the above flow mentioned.
1. Call getIndexEntry
2. Offset index is successfully cached
3. But TimeIndex received from remote storage is corrupted
4.  getIndexEntry throws Corrupt Exception
5. Validate offset index remote cache file exists
6. This time remote storage does not return corrupted time index 
7. Call getIndexEntry 
8. it should succeed , and number of remote storage calls should be
    1. Offset index  ( 1 for initial getIndexEntry)
    2. TimeIndex (2)
    3. TransactionIndex(1 , as it was not called in the 1st scenario because of exception)

iit2009060 replied:
@showuon  This use case is already covered in testIndexFileAlreadyExistOnDiskButNotInCache  , the only difference is files was created manually rather than the above flow ? Do you still see a need to merge this ?

showuon replied:
Yes, I think test the real flow makes more sense. We can change the `testIndexFileAlreadyExistOnDiskButNotInCache` into the flow above. WDYT?

iit2009060 replied:
@showuon  I can write another test case for the same , as  testIndexFileAlreadyExistOnDiskButNotInCache is covering all  types of index files existence .
The above scenario will be reproduce only in case of exception thrown  after offsetindex file is created i.e timeindex corruption or any other exception. WDYT ?

iit2009060 replied:
@showuon  please review added the test case discussed above 
**testMultipleIndexEntriesExecutionInCorruptException**

showuon reviewed (commented):
@iit2009060 , thanks for the PR. Left some comments.

iit2009060 committed:
Parametrized all corrupt index file cases

iit2009060 committed:
Adding test case of reproducing corrupt exception using code flow

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -524,23 +527,241 @@ class RemoteIndexCacheTest {
+    // validate cache entry for the above key should be null
+    assertNull(cache.internalCache().getIfPresent(rlsMetadata.remoteLogSegmentId().id()))
+    cache.getIndexEntry(rlsMetadata)
+    // Index  Files already exist
+    // rsm should not be called again
+    // instead files exist on disk
+    // should be used
```
Please make these 4 lines into 2.

iit2009060 replied:
done

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -524,23 +527,241 @@ class RemoteIndexCacheTest {
+      })
+    cache.getIndexEntry(rlsMetadata)
+    // No exception should occur
+    // Offset rsm 0,TimeIndex 1 , TxnIndex 1
```
What is `offset rsm`? OffsetIndex?

iit2009060 replied:
Updated the description. Yes it was OffsetIndex

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -524,23 +527,241 @@ class RemoteIndexCacheTest {
+    assertNull(cache.internalCache().getIfPresent(rlsMetadata.remoteLogSegmentId().id()))
+    verifyFetchIndexInvocation(1, Seq(IndexType.OFFSET, IndexType.TIMESTAMP))
+    verifyFetchIndexInvocation(0, Seq(IndexType.TRANSACTION))
+    // However cache fetch failed
+    // but it has  already created offset and time index file in remote cache dir.
+    // Current status
+    // ( cache  is null)
+    // RemoteCacheDir contain
+    // 1. Offset Index File which is fine and not corrupted
+    // 2. Time Index File which is corrupted
+    // What should be the code flow in next execution
+    // 1. No rsm call for fetching OffSet Index File.
+    // 2. Time index file should be fetched from remote storage again as it is corrupted in the first execution.
```
Please remove addition space between words.

iit2009060 replied:
done

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -524,23 +527,241 @@ class RemoteIndexCacheTest {
+    // Current status
+    // ( cache  is null)
+    // RemoteCacheDir contain
+    // 1. Offset Index File which is fine and not corrupted
+    // 2. Time Index File which is corrupted
```
The word `which` can be removed:
    // 1. Offset Index File is fine and not corrupted
    // 2. Time Index File is corrupted

iit2009060 replied:
done

iit2009060 committed:
correcting spaces and wording

divijvaidya commented:
@showuon @iit2009060  please don't wait for my review on this one. I might not get to it until next week.

iit2009060 commented:
> @showuon @iit2009060 please don't wait for my review on this one. I might not get to it until next week.

@showuon  Can you review and address @divijvaidya  comments. I have taken care of all  comments.

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -524,23 +527,239 @@ class RemoteIndexCacheTest {
+  @ParameterizedTest
+  @EnumSource(value = classOf[IndexType], names = Array("OFFSET", "TIMESTAMP", "TRANSACTION"))
+  def testCorruptCacheIndexFileExistsButNotInCache(indexType: IndexType): Unit = {
+    // create Corrupt  Index File in remote index cache
```
// create Corrupted Index File in remote index cache

showuon replied:
format is wrong:
```
    if (indexType == IndexType.OFFSET) {
      createCorruptOffsetIndexFile(cache.cacheDir())
    } else if (indexType == IndexType.TIMESTAMP) {
      createCorruptTimeIndexOffsetFile(cache.cacheDir())
    } else if (indexType == IndexType.TRANSACTION) {
      createCorruptTxnIndexForSegmentMetadata(cache.cacheDir(), rlsMetadata)
    }
```

iit2009060 replied:
done

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -524,23 +527,239 @@ class RemoteIndexCacheTest {
     // assert that parent directory for the index files is correct
     assertEquals(RemoteIndexCache.DIR_NAME, offsetIndexFile.getParent.getFileName.toString,
-      s"offsetIndex=$offsetIndexFile is not overwrite under incorrect parent")
+      s"offsetIndex=entry.offsetIndex().file().toPath is created under incorrect parent")
```
Should be `s"offsetIndex=$offsetIndexFile is created under incorrect parent")` ?

iit2009060 replied:
done

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -524,23 +527,239 @@ class RemoteIndexCacheTest {
+        val timeIdx = createTimeIndexForSegmentMetadata(metadata)
+        val txnIdx = createTxIndexForSegmentMetadata(metadata)
+        maybeAppendIndexEntries(offsetIdx, timeIdx)
+        // Create corrupt index file which would be returned by rsm
```
// Create corrupt index file which would be returned by rsm
->
 // Create corrupted index file

iit2009060 replied:
done

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -524,23 +527,239 @@ class RemoteIndexCacheTest {
+    assertNull(cache.internalCache().getIfPresent(rlsMetadata.remoteLogSegmentId().id()))
+    verifyFetchIndexInvocation(1, Seq(IndexType.OFFSET, IndexType.TIMESTAMP))
+    verifyFetchIndexInvocation(0, Seq(IndexType.TRANSACTION))
+    // However cache fetch failed
+    // but it has already created offset and time index file in remote cache dir.
```
These 2 lines can be removed.

iit2009060 replied:
done

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -524,23 +527,239 @@ class RemoteIndexCacheTest {
+    // 2. Time Index File is corrupted
+    // What should be the code flow in next execution
+    // 1. No rsm call for fetching OffSet Index File.
+    // 2. Time index file should be fetched from remote storage again as it is corrupted in the first execution.
```
Should add 3rd point:
3. Transaction index file should be fetched from remote storage

iit2009060 replied:
done

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -524,23 +527,239 @@ class RemoteIndexCacheTest {
+        }
+      })
+    cache.getIndexEntry(rlsMetadata)
+    // No exception should occur
+    // rsm should not be called for offset Index
```
`// No exception should occur` <- should be removed
`// rsm should not be called for offset Index` -> `// rsm should not be called to fetch offset Index`

iit2009060 replied:
done

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -524,23 +527,239 @@ class RemoteIndexCacheTest {
+    // validate cache entry for the above key should be null
+    assertNull(cache.internalCache().getIfPresent(rlsMetadata.remoteLogSegmentId().id()))
+    cache.getIndexEntry(rlsMetadata)
+    // Index  Files already exist ,rsm should not be called again
+    // instead files exist on disk should be used
```
// Index Files already exist, rsm should not fetch them again.

and remove L654

iit2009060 replied:
done

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -524,23 +527,239 @@ class RemoteIndexCacheTest {
+        var txnIdx = createTxIndexForSegmentMetadata(metadata)
+        maybeAppendIndexEntries(offsetIdx, timeIdx)
+        // Create corrupt index file return from RSM
+        if (testIndexType == IndexType.OFFSET) {
+          createCorruptOffsetIndexFile(tpDir)
+        }
+        else if (testIndexType == IndexType.TIMESTAMP) {
+          createCorruptTimeIndexOffsetFile(tpDir)
+        }
+        else if (testIndexType == IndexType.TRANSACTION) {
+          txnIdx = createCorruptTxnIndexForSegmentMetadata(tpDir, rlsMetadata)
+        }
```
format. Thanks.

showuon replied:
And since these logic repeat again, we can extract them into another helper method. Ex:
```
private void createCorruptedIndexFile(IndexType type) {
          if (testIndexType == IndexType.OFFSET) {
          createCorruptOffsetIndexFile(tpDir)
.....
}
```

Then, in the test, we can just call `createCorruptedIndexFile(testIndexType)` to do it.

iit2009060 replied:
done

showuon reviewed (commented):
@iit2009060 , overall LGTM, left some refactor comments.

iit2009060 committed:
Moving corrupt index file creation into a generic function

iit2009060 commented:
> @iit2009060 , overall LGTM, left some refactor comments.

@showuon  done as mentioned.

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -524,23 +527,221 @@ class RemoteIndexCacheTest {
+  @EnumSource(value = classOf[IndexType], names = Array("OFFSET", "TIMESTAMP", "TRANSACTION"))
+  def testCorruptCacheIndexFileExistsButNotInCache(indexType: IndexType): Unit = {
+    // create Corrupted Index File in remote index cache
+    createCorruptedIndexFile(indexType,cache.cacheDir())
```
nit: space after comma `(indexType, cache.cacheDir()`

iit2009060 replied:
done

showuon replied:
Thanks for the quick fix. Let's wait for the CI build completes.

iit2009060 replied:
@showuon  Unrelated test failures.
https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14482/9/tests

showuon reviewed (approved):
LGTM! Just a minor formatting comment.

iit2009060 committed:
correcting spaces and wording

kamalcph commented on storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java:
```diff
@@ -166,6 +166,11 @@ public Cache<Uuid, Entry> internalCache() {
     }
 
+    // Visible for testing
+    public File cacheDir() {
```
Can we file a ticket to move the RemoteIndexCacheTest to `storage` module under the same package?

iit2009060 replied:
@kamalcph  I have just started understanding of the kafka ecosystem , Can you help me understand the rational behind it , Then i can create a ticket with details.

kamalcph replied:
You can find more details about the intention to move the source code to Java in KAFKA-14524

kamalcph reviewed (commented):
Since we are adding new tests, can we create a new RemoteIndexCacheTest under `storage` module and start writing the tests in `java`?

The source code is in Java but the test is in scala. We can move the existing tests in a separate PR.

kamalcph commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -524,23 +527,221 @@ class RemoteIndexCacheTest {
+    // create Corrupted Index File in remote index cache
+    createCorruptedIndexFile(indexType, cache.cacheDir())
     val entry = cache.getIndexEntry(rlsMetadata)
     // Test would fail if it throws corrupt Exception
```
Can we update this comment to?

```
Test would fail it it throws exception other than CorruptIndexException
```

iit2009060 replied:
done

iit2009060 commented:
> Since we are adding new tests, can we create a new RemoteIndexCacheTest under `storage` module and start writing the tests in `java`?
> 
> The source code is in Java but the test is in scala. We can move the existing tests in a separate PR.

@kamalcph  Can i create a ticket for same and work on it after PR merge.  This has been pending from last week and already approved. 

> Since we are adding new tests, can we create a new RemoteIndexCacheTest under `storage` module and start writing the tests in `java`?
> 
> The source code is in Java but the test is in scala. We can move the existing tests in a separate PR.

kamalcph commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -554,23 +557,221 @@ class RemoteIndexCacheTest {
+    // rsm should not be called to fetch offset Index
+    verifyFetchIndexInvocation(0, Seq(IndexType.OFFSET))
+    verifyFetchIndexInvocation(1, Seq(IndexType.TIMESTAMP))
+    verifyFetchIndexInvocation(1, Seq(IndexType.TRANSACTION))
```
Txn index file was not corrupted. Can we add a comment to explain why are we fetching it again?

iit2009060 replied:
done

iit2009060 commented:
@showuon  @kamalcph  @divijvaidya  I have rebased changes and resolved conflict on the changes merged by 
the PR https://github.com/apache/kafka/pull/14381. 
Can you please review and merge it.

kamalcph reviewed (approved):
Changes in the patch, LGTM. We can rewrite it to Java in a separate patch.

showuon commented:
@iit2009060 , there are compilation failure, could you fix it? 
https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14482/10/pipeline/

iit2009060 committed:
Added comments and description

iit2009060 commented:
> @iit2009060 , there are compilation failure, could you fix it? https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14482/10/pipeline/

on it.

hudeqi commented:
> @hudeqi @showuon There is one test case failing related to the above change #14381, can you help me resolving it , in the local it is not failing ? `Build / JDK 8 and Scala 2.12 / testClearCacheAndIndexFilesWhenResizeCache() – kafka.log.remote.RemoteIndexCacheTest` `java.io.UncheckedIOException: java.nio.file.NoSuchFileException: /tmp/kafka-RemoteIndexCacheTest5502978390891743756/mwbAzcjOTtKRIDxlQrgTTw:foo-0/remote-log-index-cache/0_xg9JTaPnS4C47HfCgv2_gA.txnindex.deleted` https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14482/12/tests/

hi, I think this may be the reason: when executing `getIndexFileFromRemoteCacheDir`, the .deleted file will be listed in the `filter(Files.isRegularFile(_))` stage, but when executing `path.getFileName.toString`, the .deleted file has already been deleted because `cleanStarted` is already true to start cleaning.
If possible, can I fix this issue in your branch? @iit2009060

iit2009060 commented:
> > @hudeqi @showuon There is one test case failing related to the above change #14381, can you help me resolving it , in the local it is not failing ? `Build / JDK 8 and Scala 2.12 / testClearCacheAndIndexFilesWhenResizeCache() – kafka.log.remote.RemoteIndexCacheTest` `java.io.UncheckedIOException: java.nio.file.NoSuchFileException: /tmp/kafka-RemoteIndexCacheTest5502978390891743756/mwbAzcjOTtKRIDxlQrgTTw:foo-0/remote-log-index-cache/0_xg9JTaPnS4C47HfCgv2_gA.txnindex.deleted` https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14482/12/tests/
> 
> hi, I think this may be the reason: when executing `getIndexFileFromRemoteCacheDir`, the .deleted file will be listed in the `filter(Files.isRegularFile(_))` stage, but when executing `path.getFileName.toString`, the .deleted file has already been deleted because `cleanStarted` is already true to start cleaning. If possible, can I fix this issue in your branch? @iit2009060

Yes sure @hudeqi

Commit:
fix flaky test in RemoteIndexCacheTest

hudeqi commented:
> > > @hudeqi @showuon There is one test case failing related to the above change #14381, can you help me resolving it , in the local it is not failing ? `Build / JDK 8 and Scala 2.12 / testClearCacheAndIndexFilesWhenResizeCache() – kafka.log.remote.RemoteIndexCacheTest` `java.io.UncheckedIOException: java.nio.file.NoSuchFileException: /tmp/kafka-RemoteIndexCacheTest5502978390891743756/mwbAzcjOTtKRIDxlQrgTTw:foo-0/remote-log-index-cache/0_xg9JTaPnS4C47HfCgv2_gA.txnindex.deleted` https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14482/12/tests/
> > 
> > 
> > hi, I think this may be the reason: when executing `getIndexFileFromRemoteCacheDir`, the .deleted file will be listed in the `filter(Files.isRegularFile(_))` stage, but when executing `path.getFileName.toString`, the .deleted file has already been deleted because `cleanStarted` is already true to start cleaning. If possible, can I fix this issue in your branch? @iit2009060
> 
> Yes sure @hudeqi . This seems to be a case of flaky behaviour i.e. race condition ?

I think so, please cherry-pick this commit, thanks. @iit2009060 

> > > @hudeqi @showuon There is one test case failing related to the above change #14381, can you help me resolving it , in the local it is not failing ? `Build / JDK 8 and Scala 2.12 / testClearCacheAndIndexFilesWhenResizeCache() – kafka.log.remote.RemoteIndexCacheTest` `java.io.UncheckedIOException: java.nio.file.NoSuchFileException: /tmp/kafka-RemoteIndexCacheTest5502978390891743756/mwbAzcjOTtKRIDxlQrgTTw:foo-0/remote-log-index-cache/0_xg9JTaPnS4C47HfCgv2_gA.txnindex.deleted` https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14482/12/tests/
> > 
> > 
> > hi, I think this may be the reason: when executing `getIndexFileFromRemoteCacheDir`, the .deleted file will be listed in the `filter(Files.isRegularFile(_))` stage, but when executing `path.getFileName.toString`, the .deleted file has already been deleted because `cleanStarted` is already true to start cleaning. If possible, can I fix this issue in your branch? @iit2009060
> 
> Yes sure @hudeqi . This seems to be a case of flaky behaviour i.e. race condition ?

I think so, please cherry-pick this commit: https://github.com/hudeqi/kafka/commit/854c773e708a8352d018986873743b325e97c3f5, thanks. @iit2009060

iit2009060 commented:
> > > > @hudeqi @showuon There is one test case failing related to the above change #14381, can you help me resolving it , in the local it is not failing ? `Build / JDK 8 and Scala 2.12 / testClearCacheAndIndexFilesWhenResizeCache() – kafka.log.remote.RemoteIndexCacheTest` `java.io.UncheckedIOException: java.nio.file.NoSuchFileException: /tmp/kafka-RemoteIndexCacheTest5502978390891743756/mwbAzcjOTtKRIDxlQrgTTw:foo-0/remote-log-index-cache/0_xg9JTaPnS4C47HfCgv2_gA.txnindex.deleted` https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14482/12/tests/
> > > 
> > > 
> > > hi, I think this may be the reason: when executing `getIndexFileFromRemoteCacheDir`, the .deleted file will be listed in the `filter(Files.isRegularFile(_))` stage, but when executing `path.getFileName.toString`, the .deleted file has already been deleted because `cleanStarted` is already true to start cleaning. If possible, can I fix this issue in your branch? @iit2009060
> > 
> > 
> > Yes sure @hudeqi . This seems to be a case of flaky behaviour i.e. race condition ?
> 
> I think so, please cherry-pick this commit: [hudeqi@854c773](https://github.com/hudeqi/kafka/commit/854c773e708a8352d018986873743b325e97c3f5), thanks. @iit2009060

@hudeqi  This is given me bad object/revision . Can you have the command handy to do cherry pick from another fork ?
git cherry-pick 854c773e708a8352d018986873743b325e97c3f5
fatal: bad object 854c773e708a8352d018986873743b325e97c3f5
This seems to be existing in your fork repository

iit2009060 commented:
> > > > > @hudeqi @showuon There is one test case failing related to the above change #14381, can you help me resolving it , in the local it is not failing ? `Build / JDK 8 and Scala 2.12 / testClearCacheAndIndexFilesWhenResizeCache() – kafka.log.remote.RemoteIndexCacheTest` `java.io.UncheckedIOException: java.nio.file.NoSuchFileException: /tmp/kafka-RemoteIndexCacheTest5502978390891743756/mwbAzcjOTtKRIDxlQrgTTw:foo-0/remote-log-index-cache/0_xg9JTaPnS4C47HfCgv2_gA.txnindex.deleted` https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14482/12/tests/
> > > > 
> > > > 
> > > > hi, I think this may be the reason: when executing `getIndexFileFromRemoteCacheDir`, the .deleted file will be listed in the `filter(Files.isRegularFile(_))` stage, but when executing `path.getFileName.toString`, the .deleted file has already been deleted because `cleanStarted` is already true to start cleaning. If possible, can I fix this issue in your branch? @iit2009060
> > > 
> > > 
> > > Yes sure @hudeqi . This seems to be a case of flaky behaviour i.e. race condition ?
> > 
> > 
> > I think so, please cherry-pick this commit: [hudeqi@854c773](https://github.com/hudeqi/kafka/commit/854c773e708a8352d018986873743b325e97c3f5), thanks. @iit2009060
> 
> @hudeqi This is given me bad object/revision . Can you have the command handy to do cherry pick from another fork ? git cherry-pick [854c773](https://github.com/apache/kafka/commit/854c773e708a8352d018986873743b325e97c3f5) fatal: bad object [854c773](https://github.com/apache/kafka/commit/854c773e708a8352d018986873743b325e97c3f5) This seems to be existing in your fork repository I figured out through [this](https://simonhartcher.com/how-to-cherry-pick-a-git-commit-from-a-fork) link. Will update soon the PR with changes.

@hudeqi  Please review it

iit2009060 commented:
Unrelated test failures 
https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14482/14/tests
@showuon  @divijvaidya  can we merge this changes ?

iit2009060 commented:
@showuon  Can we merge this changes ?

showuon commented on core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala:
```diff
@@ -540,6 +545,8 @@ class RemoteIndexCacheTest {
       "Failed to mark cache entry for cleanup after resizing cache.")
     TestUtils.waitUntilTrue(() => cacheEntry.isCleanStarted,
       "Failed to cleanup cache entry after resizing cache.")
+    TestUtils.waitUntilTrue(() => cacheEntry.isCleanFinished,
+      "Failed to finish cleanup cache entry after resizing cache.")
 
     // verify no index files on remote cache dir
     TestUtils.waitUntilTrue(() => !getIndexFileFromRemoteCacheDir(LogFileUtils.INDEX_FILE_SUFFIX).isPresent,
```
@iit2009060 @hudeqi , adding an `isCleanFinished` flag just for test is not a good solution. In this case, could we catch the exception in the `getIndexFileFromRemoteCacheDir`?

iit2009060 replied:
@hudeqi  Can you resolve the above comment ?

iit2009060 replied:
@showuon  I have removed the `isCleanFinished` flag changes and catch the `Exception` in the method.
cc @hudeqi 
Please review it .

iit2009060 committed:
Catch FileNotFoundException in resize cache test

showuon reviewed (approved):
LGTM! Thanks for the update.

showuon commented:
Failed tests are unrelated and also failed in trunk build.
